types/set.go

```diff
@@ -0,0 +1,276 @@
+	// elements that are in the set are known.
+	Unknown bool
+
+	// Null will be set to true if the set is null, either because it was
+	// omitted from the configuration, state, or plan, or because it was
+	// explicitly set to null.
+	Null bool
+
+	// Elems are the elements in the set.
+	Elems []attr.Value
+
+	// ElemType is the tftypes.Type of the elements in the set. All
+	// elements in the set must be of this type.
+	ElemType attr.Type
+}
+
+// ElementsAs populates `target` with the elements of the Set, throwing an
+// error if the elements cannot be stored in `target`.
+func (s Set) ElementsAs(ctx context.Context, target interface{}, allowUnhandled bool) diag.Diagnostics {
+	// we need a tftypes.Value for this Set to be able to use it with our
+	// reflection code
+	values, err := s.ToTerraformValue(ctx)
+	if err != nil {
+		return diag.Diagnostics{
+			diag.NewErrorDiagnostic(
+				"Set Element Conversion Error",
+				"An unexpected error was encountered trying to convert set elements. This is always an error in the provider. Please report the following to the provider developer:\n\n"+err.Error(),
+			),
+		}
+	}
+	return reflect.Into(ctx, SetType{ElemType: s.ElemType}, tftypes.NewValue(tftypes.Set{
+		ElementType: s.ElemType.TerraformType(ctx),
+	}, values), target, reflect.Options{
+		UnhandledNullAsEmpty:    allowUnhandled,
+		UnhandledUnknownAsEmpty: allowUnhandled,
+	})
+}
+
+// Type returns a SetType with the same element type as `s`.
+func (s Set) Type(ctx context.Context) attr.Type {
+	return SetType{ElemType: s.ElemType}
+}
+
+// ToTerraformValue returns the data contained in the AttributeValue as
+// a Go type that tftypes.NewValue will accept.
+func (s Set) ToTerraformValue(ctx context.Context) (interface{}, error) {
+	if s.Unknown {
+		return tftypes.UnknownValue, nil
+	}
+	if s.Null {
+		return nil, nil
+	}
+	vals := make([]tftypes.Value, 0, len(s.Elems))
+	for _, elem := range s.Elems {
+		val, err := elem.ToTerraformValue(ctx)
+		if err != nil {
+			return nil, err
+		}
+		err = tftypes.ValidateValue(s.ElemType.TerraformType(ctx), val)
+		if err != nil {
+			return nil, fmt.Errorf("error validating terraform type: %w", err)
+		}
+		vals = append(vals, tftypes.NewValue(s.ElemType.TerraformType(ctx), val))
+	}
+	return vals, nil
+}
+
+// Equal must return true if the AttributeValue is considered
+// semantically equal to the AttributeValue passed as an argument.
+func (s Set) Equal(o attr.Value) bool {
+	other, ok := o.(Set)
+	if !ok {
+		return false
+	}
+	if s.Unknown != other.Unknown {
+		return false
+	}
+	if s.Null != other.Null {
+		return false
+	}
+	if s.ElemType == nil && other.ElemType != nil {
+		return false
+	}
+	if s.ElemType != nil && !s.ElemType.Equal(other.ElemType) {
+		return false
+	}
+	if len(s.Elems) != len(other.Elems) {
+		return false
+	}
+	for _, elem := range s.Elems {
+		if !other.contains(elem) {
+			return false
+		}
+	}
+	return true
+}
+
+func (s Set) contains(v attr.Value) bool {
```

Comment:
This seems like a useful function to export. Do we want to export it?

Reply:
Is manually working with a framework-defined `Set` a goal for our implementation or do we want to say that is an implementation detail for outside the framework? Currently developers can either build this manually via `Elems` or use reflection, similar to `List`. If we add methods for working with `Set` via framework-defined type methods, my next question would be should we offer these methods for the `List` and `Map` types for consistency?

Reply:
You're right, let's make this a follow-up enhancement and apply it to all the types.